### ISSUE TITLE
DEX integration

### ISSUE THREAD

**Issue by hurek** [barter-rs/barter-rs#27]
Hello! is it possible to use your library to trade on DEX?

**Comment by xsa-dev** [comment]
Now in barter-data there is code to work only with these exchanges:

- binance
- bitfinex
- bitmex
- bybit
- coinbase
- gateio
- kraken
- okx

If you need support for something specific, please let us know where you are going to connect. 

In addition to backtesting data, you can use pre-processed data in json format out of the box. Take a look at the examples in barter-data folder.

**Comment by xsa-dev** [comment]
To make it easier to understand and implement the developments, a similar format should be followed:

## Name: Exchange Integration Request: _Hyperliquid_

### Exchange Summary

**Exchange Name:** Hyperliquid

**URL of the Exchange:** [hyperliquid.xyz](https://hyperliquid.xyz)

**URL of the Exchange's API docs:** [Hyperliquid API Documentation](https://hyperliquid.gitbook.io/hyperliquid-docs/for-developers/api)

**Reason for Implementation:** Hyperliquid is a highly regarded decentralized exchange (DEX).

**Comment by xsa-dev** [comment]
Found[API Rust SDK](https://github.com/hyperliquid-dex/hyperliquid-rust-sdk). It might be useful for us.

**Comment by hurek** [comment]
I would like to be able to write adapters for dex exchanges. It would also be cool to be able to choose strategies (regular trading, arbitrage, sniping).

**Comment by xsa-dev** [comment]
Thank you for sharing your ideas. It's great that you're looking to expand the functionality of a project to include decentralized exchanges (DEXs) and various trading strategies.

Based on the context provided, I don't have specific information about DEX adapters or strategy implementations in the current codebase. However, I can offer some general suggestions on how you might approach this within the Barter framework:

1. DEX Adapters:
   To add support for DEX exchanges, you would likely need to create new modules similar to the existing exchange modules. You might consider:

   - Creating a new module under `barter-data/src/exchange/` for each DEX you want to support.
   - Implementing the necessary traits and structures to interact with the DEX's API or smart contracts.
   - Ensuring that the DEX adapter conforms to the existing `ExchangeWsStream` or similar interfaces for consistency.

2. Trading Strategies:
   For implementing different trading strategies, you could:

   - Extend the existing `Strategy` or `SignalGenerator` trait (mentioned in `barter/src/lib.rs`) to accommodate various strategy types.
   - Create separate modules for each strategy type (e.g., regular trading, arbitrage, sniping) under a new `strategies/` directory.
   - Implement the strategy-specific logic within these new modules.

Here's a conceptual example of how you might structure this:

```rust
// In a new file, e.g., barter-strategy/src/lib.rs
pub mod regular;
pub mod arbitrage;
pub mod sniping;

pub trait Strategy {
    fn generate_signal(&self, market_data: &MarketData) -> Signal;
}

// In barter-strategy/src/regular.rs
pub struct RegularTradingStrategy;

impl Strategy for RegularTradingStrategy {
    fn generate_signal(&self, market_data: &MarketData) -> Signal {
        // Implement regular trading logic
    }
}

// Similar implementations for arbitrage and sniping strategies
```

To make strategies selectable, you could use an enum or a trait object:

```rust
pub enum StrategyType {
    Regular(RegularTradingStrategy),
    Arbitrage(ArbitrageStrategy),
    Sniping(SnipingStrategy),
}

impl Strategy for StrategyType {
    fn generate_signal(&self, market_data: &MarketData) -> Signal {
        match self {
            StrategyType::Regular(s) => s.generate_signal(market_data),
            StrategyType::Arbitrage(s) => s.generate_signal(market_data),
            StrategyType::Sniping(s) => s.generate_signal(market_data),
        }
    }
}
```

Remember, these are just conceptual examples. The actual implementation would need to be integrated with our existing codebase and might require adjustments to fit within the current architecture.

To proceed, I'd recommend:

1. Designing the interfaces for DEX adapters and new strategy types.
2. Implementing a proof-of-concept for one DEX and one new strategy type.
3. Integrating these with your existing `Trader` and `Engine` components.
4. Iterating and expanding to cover more DEXs and strategies.

**Comment by hurek** [comment]
I have quite a lot of experience with dex exchanges on solana. But I have a lot of questions how to integrate all this into the existing barter-rs architecture. 

For example, we can use default solana websocket or GRPC geyser streams to get data stream. 
From most DEX exchanges we can't get OHLC (this data is paid and comes with a long delay). This begs the question - how to effectively aggregate trade tick into ohlc/bars etc. 

The next aspect to note is that for sniping/arbitrage strategies, we do not hardcode a trader instance with the parameters of a particular AMM pool. Essentially, the lifecycle of sniping is: 
1) detecting new AMM pools 
2) optional security checks 
3) sending a buy transaction (with or without tip) 
4) listening to events and having the strategy determine a sell signal. 

As for atomic arbitrage (within a single transaction), the task here is usually reduced to 
1) Initial loading and filtering of all suitable AMM/CLMM pools and creating a MarketGraph 
2) Listening to transaction events, determining the arbitrage opportunity
3) Sending the transaction

Common trading - just common trading. market, limit orders, etc.

Basically, these 3 strategies will work for any network (not just solana). But at the same time, we have to implement adapters not only for each network at the infrastructure level, but also for each necessary DEX within that network.

I would be interested in talking to you about the architecture and integration of these features into barter-rs. I might be able to contribute.

**Comment by xsa-dev** [comment]
On behalf of xsa-dev, I'd like to invite you to join our barter-rs community on Discord. We would be delighted to discuss your ideas about DEX integration and various trading strategies in more detail. Our Discord community is an excellent place for exchanging thoughts, getting feedback, and collaborating on the project's development. There, you'll be able to directly communicate with developers, including myself, and other community members.

Here's the link to join: https://discord.gg/wE7RqhnQMV

We look forward to seeing you there and continuing the discussion of your proposals in a more interactive format. Your experience with DEX on Solana could be precious for the project.

**Comment by shali1995** [comment]
> I have quite a lot of experience with dex exchanges on solana. But I have a lot of questions how to integrate all this into the existing barter-rs architecture.
> 
> For example, we can use default solana websocket or GRPC geyser streams to get data stream. From most DEX exchanges we can't get OHLC (this data is paid and comes with a long delay). This begs the question - how to effectively aggregate trade tick into ohlc/bars etc.
> 
> The next aspect to note is that for sniping/arbitrage strategies, we do not hardcode a trader instance with the parameters of a particular AMM pool. Essentially, the lifecycle of sniping is:
> 
> 1. detecting new AMM pools
> 2. optional security checks
> 3. sending a buy transaction (with or without tip)
> 4. listening to events and having the strategy determine a sell signal.
> 
> As for atomic arbitrage (within a single transaction), the task here is usually reduced to
> 
> 1. Initial loading and filtering of all suitable AMM/CLMM pools and creating a MarketGraph
> 2. Listening to transaction events, determining the arbitrage opportunity
> 3. Sending the transaction
> 
> Common trading - just common trading. market, limit orders, etc.
> 
> Basically, these 3 strategies will work for any network (not just solana). But at the same time, we have to implement adapters not only for each network at the infrastructure level, but also for each necessary DEX within that network.
> 
> I would be interested in talking to you about the architecture and integration of these features into barter-rs. I might be able to contribute.

how can we talk, do you have telegram?
